### PR TITLE
Remove middle wildcard support in rules.js

### PR DESCRIPTION
**Pull request by cschanaj**
This replaces #17669 to drop support for middle wildcards in the runtime. hopefully #17711

### CODE REVIEW

**Comment by pipboy96** [comment]
@cschanaj Can you add a test case that checks if middle wildcard causes error and gets ignored as it's supposed to?

**Comment by pipboy96** [comment]
This may not necessarily work, since some rulesets depend on `*.example.com` also matching `a.b.example.com`.

File: chromium/background-scripts/rules.js [hunk 1]
```diff
@@ -562,22 +562,23 @@ RuleSets.prototype = {
...
 
-    // Replace each portion of the domain with a * in turn
+    // Replace www.example.com with www.example.*
+    // eat away from the right for once and only once
     let segmented = host.split(".");
-    for (let i = 0; i < segmented.length; i++) {
-      let tmp = segmented[i];
-      segmented[i] = "*";
+    if (segmented.length > 1) {
+      let tmp = segmented[segmented.length - 1];
```
Comment (Hainish) [inline chromium/background-scripts/rules.js]: `const` can be used here

File: chromium/test/rules_test.js [hunk 2]
```diff
@@ -203,15 +203,23 @@ describe('rules.js', function() {
           assert.deepEqual(res3, new Set(value), 'wildcard matches sub domains');
         });
 
-        it('matches middle wildcards', function() {
-          let target = 'sub.*.' + host;
+        it('matches right wildcards', function() {
+          let target = host + '.*';
```
Comment (Hainish) [inline chromium/test/rules_test.js]: `const` can be used here

File: chromium/test/rules_test.js [hunk 3]
```diff
@@ -203,15 +203,23 @@ describe('rules.js', function() {
...
           this.rsets.targets.set(target, value);
 
-          let res1 = this.rsets.potentiallyApplicableRulesets('sub.star.' + host);
+          let res1 = this.rsets.potentiallyApplicableRulesets(host + '.tld');
```
Comment (Hainish) [inline chromium/test/rules_test.js]: `const` can be used here

File: chromium/test/rules_test.js [hunk 4]
```diff
@@ -203,15 +203,23 @@ describe('rules.js', function() {
...
           assert.deepEqual(res1, new Set(value), 'default case');
 
-          let res2 = this.rsets.potentiallyApplicableRulesets('sub.foo.bar.' + host);
-          assert.isEmpty(res2, new Set(value), 'only matches one label');
+          let res2 = this.rsets.potentiallyApplicableRulesets(host + '.tld.com');
```
Comment (Hainish) [inline chromium/test/rules_test.js]: `const` can be used here

File: chromium/test/rules_test.js [hunk 5]
```diff
@@ -203,15 +203,23 @@ describe('rules.js', function() {
...
+          assert.isEmpty(res2, 'wildcard matches second level domains');
+        });
+
+        it('ignore middle wildcards', function() {
+          let target = 'www.*.' + host;
```
Comment (Hainish) [inline chromium/test/rules_test.js]: `const` can be used here

File: chromium/test/rules_test.js [hunk 6]
```diff
@@ -203,15 +203,23 @@ describe('rules.js', function() {
...
+          this.rsets.targets.set(target, value);
+
+          let res1 = this.rsets.potentiallyApplicableRulesets('www.cdn.' + host);
```
Comment (Hainish) [inline chromium/test/rules_test.js]: `const` can be used here

**Pull request merged**